src/service_discovery/mod.rs

```diff
@@ -112,15 +107,11 @@ impl ServiceDiscovery {
 
     fn read(&mut self, core: &mut Core, poll: &Poll) {
         loop {
-            match self.socket.recv_from(&mut self.read_buf) {
-                Ok((bytes_rxd, peer_addr)) => {
-                    self.handle_incoming_msg(bytes_rxd, peer_addr, core, poll);
-                }
-                Err(ref e)
-                    if e.kind() == ErrorKind::Interrupted || e.kind() == ErrorKind::WouldBlock =>
-                {
-                    return
+            match self.socket.read_frm() {
+                Ok(Some((msg, peer_addr))) => {
+                    self.handle_incoming_msg(msg, peer_addr);
                 }
+                Ok(None) => return,
                 Err(e) => {
```

Comment:
hmm this is not correct anymore after your changes above. The deserialisation error due to 1 peer will now cause the collapse of entire ServDisc server and affect all other peers. So filter that separately and continue the loop in that case to cater to others.

Reply:
yeah, you're right. I'll fix this.